lib/AST/AST.h

```diff
@@ -206,6 +206,21 @@ class ASTImpl : public std::enable_shared_from_this<ASTImpl> {
   // contexts of tokens with macro roles.
   void LinkMacroTokenContexts(void);
 
+  // Returns the index of the next final expansion or file token in this AST's
+  // token list
+  inline std::optional<uint64_t>
+      NextFinalExpansionOrFileTokenIndex(const Token &tok) {
```

Comment:
Deindent.

Reply:
Move implementation to `.cpp`.

Reply:
Logically I think this should be a method on `TokenImpl`. You may also want to consider making this return a `std::optional<Token>` if that would make using this API easier.

Reply:
Ok, I changed the return type to `Token` and that is preferable. However I made this method a method of `Token` instead of `TokenImpl` because that was a bit easier to work with (`Token`s already have an `ASTImpl` member, their constructor involving `TokenImpl` is private, etc.).